[docs] Add a Linking section (#30958)
expo/expo: docs/pages/linking/into-your-app.mdx

---
title: Linking into your app
sidebar_title: Into your app

> description: Learn how to handle an incoming URL in your app

Author: ```suggestion
description: Learn how to handle an incoming URL in your app by creating a deep link.
```

> description: Learn how to open and handle a URL from other apps based on the scheme for your by creating a deep link.

Reviewer: ```suggestion
description: Learn how to handle an incoming URL in your app
```
Realizing maybe there was some search optimization happening here, so feel free to add back words

  return <Text>URL: {url}</Text>;
}
```

The `Linking.useURL()` hook works behind the scenes by following these imperative methods:

- The link that launched the app is initially returned using [`Linking.getInitialURL()`](/versions/latest/sdk/linking/#linkinggetinitialurl)
- Any new links triggered while the app is already open are observed with [`Linking.addEventListener('url', callback)`](/versions/latest/sdk/linking/#linkingaddeventlistenertype-handler)

## Parse URLs

You can use the [`Linking.parse()`](/versions/latest/sdk/linking/#linkingparseurl) method to parse the **path**, **hostname**, and **query parameters** from a URL. This method extracts deep linking information and considers nonstandard implementations.

```tsx index.tsx|collapseHeight=440
import * as Linking from 'expo-linking';

export default function Home() {
  const url = Linking.useURL();

  if (url) {
    const { hostname, path, queryParams } = Linking.parse(url);

    console.log(
      `Linked to app with hostname: ${hostname}, path: ${path} and data: ${JSON.stringify(
        queryParams
      )}`
    );
  }

  return (
    /* @hide Your React component here. */ /* @end */
  )
}
```

## Limitations

> If a user does not have your app installed, deep links don't link to anything. Attribution services like [Branch](https://www.branch.io/deep-linking/) offer solutions for condistionally linking to your app or web page.

Reviewer: ```suggestion
If a user does not have your app installed, deep links to your app will not work. Attribution services like [Branch](https://www.branch.io/deep-linking/) offer solutions for condistionally linking to your app or web page.
```

> Universal links is another solution you can use to handle such cases, in which the route to your web page determins when the app is installed or not. For more details, see [Universal Links](/linking/ios-universal-links/).

Reviewer: ```suggestion
Universal links is another solution you can use to handle such cases, in which the route to your web page determines whether or not the app is installed. For more details, see [Universal Links](/linking/ios-universal-links/).
```

- Alternatively, you can open the link by searching for it in the device's native browser. For example, open Safari on iOS and type `myapp://` in the address bar to prompt the device to open your app.

<Collapsible summary="Test a link using Expo Go">

By default, [Expo Go](https://expo.dev/go) uses the `exp://` scheme. If you link to `exp://` without specifying a URL address afterward, it will open the app to the home screen. In development, your app's complete URL looks like `exp://127.0.0.1:8081`.

To open the `/details` screen while testing on Expo Go, you can use `npx uri-scheme`:

<Terminal
  cmd={['$ npx uri-scheme open exp://127.0.0.1:8081/--/somepath/into/app?hello=world --ios']}
/>

In Expo Go, `/--/` is added to the URL when a path is specified. This indicates to Expo Go that the substring after it corresponds to the deep link path and is not part of the path to the app itself.

By default, `exp://` is replaced with `http://` when opening a URL in Expo Go. You can also use `exps://` to open `https://` URLs. However, `exps://` does not currently support loading sites with insecure TLS certificates.

</Collapsible>

## Handle URLs

> You can observe links that launch your app using the [`Linking.useURL()`](/versions/latest/sdk/linking/#useurl) hook from [`expo-linking`](/versions/latest/sdk/linking/).

Reviewer: might be worth a call-out to the effect of "you can ignore all this with Expo Router"

  return <Text>URL: {url}</Text>;
}
```

The `Linking.useURL()` hook works behind the scenes by following these imperative methods:

- The link that launched the app is initially returned using [`Linking.getInitialURL()`](/versions/latest/sdk/linking/#linkinggetinitialurl)
- Any new links triggered while the app is already open are observed with [`Linking.addEventListener('url', callback)`](/versions/latest/sdk/linking/#linkingaddeventlistenertype-handler)

## Parse URLs

You can use the [`Linking.parse()`](/versions/latest/sdk/linking/#linkingparseurl) method to parse the **path**, **hostname**, and **query parameters** from a URL. This method extracts deep linking information and considers nonstandard implementations.

```tsx index.tsx|collapseHeight=440
import * as Linking from 'expo-linking';

export default function Home() {
  const url = Linking.useURL();

  if (url) {
    const { hostname, path, queryParams } = Linking.parse(url);

    console.log(
      `Linked to app with hostname: ${hostname}, path: ${path} and data: ${JSON.stringify(
        queryParams
      )}`
    );
  }

  return (
    /* @hide Your React component here. */ /* @end */
  )
}
```

> ## When to not use deep links

Reviewer: Kind of goes for the entire page, but if we update the overview to clarify a link vs deep link, might be good to sync this language, because a lot of this is talking about links that may or may not be deep.

> If a user does not has your app installed and follows a link to your app's custom scheme, the device's operating system will indicate that the page couldn't be opened without providing any further information. This results in a poor user experience, and unfortunately, there's no way to work around this in a web browser.

Reviewer: ```suggestion
If a user does not have your app installed and follows a link to your app's custom scheme, the device's operating system will indicate that the page couldn't be opened without providing any further information. This results in a poor user experience, and unfortunately, there's no way to work around this in a web browser.
```

}
```

The `Linking.useURL()` hook works behind the scenes by following these imperative methods:

- The link that launched the app is initially returned using [`Linking.getInitialURL()`](/versions/latest/sdk/linking/#linkinggetinitialurl)
- Any new links triggered while the app is already open are observed with [`Linking.addEventListener('url', callback)`](/versions/latest/sdk/linking/#linkingaddeventlistenertype-handler)

## Parse URLs

You can use the [`Linking.parse()`](/versions/latest/sdk/linking/#linkingparseurl) method to parse the **path**, **hostname**, and **query parameters** from a URL. This method extracts deep linking information and considers nonstandard implementations.

```tsx index.tsx|collapseHeight=440
import * as Linking from 'expo-linking';

export default function Home() {
  const url = Linking.useURL();

  if (url) {
    const { hostname, path, queryParams } = Linking.parse(url);

    console.log(
      `Linked to app with hostname: ${hostname}, path: ${path} and data: ${JSON.stringify(
        queryParams
      )}`
    );
  }

  return (
    /* @hide Your React component here. */ /* @end */
  )
}
```

## When to not use deep links

> If a user does not has your app installed and follows a link to your app's custom scheme, the device's operating system will indicate that the page couldn't be opened without providing any further information. This results in a poor user experience, and unfortunately, there's no way to work around this in a web browser.

Author: I'm not sure (after working on iOS universal links and Android App Links), the last statement is true anymore, since Apple Smart Banner handles this for iOS. I think I might not have complete context.

@brentvatne: is the last statement still relevant?

Reviewer: We could walk around this a bit by being less prescriptive ... instead of "when not to use deep links", this could be "limitations" and something as simple as, "If the user doesn't have your app installed, deep links don't link to anything." and optionally "Attribution services like Branch offer solutions for conditionally linking to your app or a web page" (at least as I recall) and/ or "one solution is to use universal links, which will route to your web page when the app isn't installed [learn more]"
